Content for custom html (#880)
novuhq/docs: content/docs/platform/integrations/email/writing-email-template.mdx

---
title: 'Writing Email Template'
sidebarTitle: 'Writing Email Template'
description: 'Learn how to build email templates in Novu using either the block editor or code editor with custom HTML'
icon: 'file-pen'
---

You can create email templates in Novu using the following methods:

- **Block editor**: A visual drag-and-drop editor using prebuilt content blocks and basic variables.
- **Code editor (custom HTML)**: Full HTML and [LiquidJS](https://liquidjs.com/index.html) control for layout, logic, styling, and rendering.
- **API**: Submit raw HTML and Liquid templates via the API.

## Blocks editor

The block editor is a `WYSIWYG` editor that allows you to create and edit email templates. It has two fields: `Subject` and `Body`.

- **Subject**: Title of the email. It supports variables and can be customized based on the subscriber properties and payload variables.
- **Body**: Main content of the email. It is made of blocks.

> The block editor body is made of blocks. A block can be added by clicking on the plus (+) icon in the top-left corner of the editor or by adding a forward slash (`/`). In both cases, a menu will appear with the list of supported blocks. Click on the desired block to add it to the editor. The menu option next to the plus (+) icon can be used to **duplicate** or **delete** the block.

Reviewer: ```suggestion
The block editor body is made of blocks. A block can be added by clicking on the plus (+) icon in the top-left corner of the editor or by adding a forward slash (`/`). In both cases, a menu appears with the list of supported blocks. Click on the desired block to add it to the editor. The menu option next to the plus (+) icon can be used to **duplicate** or **delete** the block.
```

- **Numbered List**: Numbered list (numeric digits like 1,2,3)
- **Image**: Full widh image with absolute url, image position can be customized
- **Section**: Create a section to group content together
- **Column**: Creates columns to group content together, useful for responsive design
- **Divider**: Separates the content, adds a line to highlight the separation
- **Spacer**: Spacer to add space between two blocks (available in sm, lg and xl sizes)
- **Button**: Call to action button to link to a page or URL that can be customized with text, URL, color, and size and background color.
- **Hard Break**: Adds a line break
- **Blockquote**: Adds a blockquote
- **Custom HTML code**: Insert raw HTML with variable support
- **Repeat**: Can be used for interation on array of data
- **Show**: Can be used to conditionally show content based on a condition, use eye icon to toggle the visibility of the content

### Custom HTML block

The Custom HTML block allows you to insert raw HTML inside a block editor layout.

- Supports basic HTML structure.
- Allows injecting variables using `{{ ... }}`.
- Does not support LiquidJS control structures like `{% for %}` or `{% if %}`.

> Use this block when you need inline markup control within a visual layout, but don't require full Liquid logic.

Reviewer: Liquid or LiquidJS?

  alt="Show block"
/>

In the above image example, the Track your order button is displayed only if a subscriber has the showTracking field set to true in their custom data attribute. Similarly, you can use payload variables to show content conditionally.

## Code editor (custom HTML)

> The code editor provides a built-in interface for writing raw HTML and embedding Liquid expressions directly inside the Novu dashboard. This gives you full control over layout, structure, and rendering beyond the capabilities of Novu's block editor.

Reviewer: Why do you use "Liquid" in some cases and "LiquidJS" in others? Be consistent.

/>

In above example, if subscriber has `showTracking` field set to `true` in custom `data` attribute, then the `Track your order` button will be shown in the email sent to the subscriber. Similarly, payload variables can be used to conditionally show the content.

## Code editor (custom HTML)

> Novu’s custom HTML for email lets you build fully custom email templates by writing raw HTML combined with full LiquidJS support for dynamic data binding, control flow, and personalization. This gives you full control over layout, structure, and rendering, beyond the capabilities of Novu's block editor.

Reviewer 2: `Novu’s custom HTML for email lets` this beginning sound weird
```suggestion
Custom HTML Editor lets you build email templates by writing raw HTML combined with full LiquidJS support for dynamic data binding, control flow, and personalization. This gives you full control over layout, structure, and rendering, beyond the capabilities of Novu's block editor.
```
or smth similar

This feature supports advanced use cases such as:
- Migrating existing HTML email templates.
- Implementing full design systems.
- Executing flow control logic directly inside email templates.
- Complete layout and styling freedom without editor constraints.

<Callout type='info'>Custom HTML is available for customers on the Pro tier and above.</Callout>

### Authoring custom HTML emails

You can create custom HTML emails using either the Novu dashboard (Code editor) or via API passthrough. Both methods provide full control over email rendering.

> #### Using the code editor in Novu dashboard

Reviewer 2: ```suggestion
#### Using the code editor
```

> The Code editor allows you to write raw HTML and embed LiquidJS expressions directly within the Novu dashboard.

Reviewer 2: repeated sentence from the main section and the ending is a bit off

To use it:
1. Open an email step in your workflow.
2. Select the **Code editor** tab to switch to Custom HTML mode.
3. Start writing your template using HTML and LiquidJS syntax.

<Callout type='warn'>Once you switch to the Code editor, reverting the template back to the Block editor is not supported as some HTML constructs may not map cleanly to block definitions, so further edits must be made directly in HTML.</Callout>

#### Using the API

You can also submit fully pre-rendered HTML directly via the Novu API. This allows you to control template rendering server-side and submit the complete HTML payload for delivery.

When using API passthrough:

- Novu delivers the submitted HTML exactly as provided.
- No additional wrappers, headers, or system-injected layout elements are added.

> - LiquidJS is not executed server-side; the HTML you submit must be fully rendered.

Reviewer 2: `Novu delivers the submitted HTML exactly as provided` - During the sending phase Novu compiles the HTML + Liquid template in conjunction with payload and subscriber context information and delivers the fully rendered HTML email.

`No additional wrappers, headers, or system-injected layout elements are added` not true, we include the Novu branding footer by default.

`LiquidJS is not executed server-side...` as mentioned above, we compile the template during the sending phase, so it is executed on our end, but is saved in our DB as they send it (HTML+Liquid)

> API passthrough is commonly used when you maintain your own rendering pipeline and wish to integrate Novu as a delivery mechanism.

Reviewer 2: needs to be rephrased
